p-org/P
# [PEx] Updates to P tutorial and CLI for PEx, support interface refinement in PEx, refactoring

Updates include:
[PEx]
- Renames PExplicit -> PEx, GlobalFunctions -> PExForeignCode, mode explicit -> pex
- Refactors PEx IR generator
- Add support for refinement interfaces (e.g., AbstractBank interface in ClientServer from P Tutorial)

[P CLI]
- Makes CLI options `--mode <bugfinding | pex>` and `--sch-pex <random | dfs>` visible

[P Tutorial]
- ClientServer: limits the number of withdraw requests to at most 5, adds 2 hints to limit data non-determinism
- TwoPhaseCommit: adds hint to create random transaction without foreign code and with less data non-determinism
- EspressoMachine: no change

Testing:
1. Build P with PEx by running the following: (requires Maven v3.7 or higher)
````
cd Src/PRuntimes/PExRuntime
./scripts/build.sh
````
2. Compile and run PEx on a P Tutorial example
````
cd Tutorial/3_EspressoMachine
pl compile --mode pex
pl check --mode pex -tc tcCrazyUserUsingCoffeeMachine --timeout 60
````

3. Check with expected output:
````
$ pl check --mode pex -tc tcCrazyUserUsingCoffeeMachine --timeout 60
.. Searching for a P compiled file locally in folder ./PGenerated/
.. Found a P compiled file: ./PGenerated/PEx/target/EspressoMachine-jar-with-dependencies.jar
.. Checking ./PGenerated/PEx/target/EspressoMachine-jar-with-dependencies.jar
.. Test case :: tcCrazyUserUsingCoffeeMachine
... Checker is using 'random' strategy (seed:1723671344241)
--------------------
   Time     Memory   Step    Schedule   Timelines         Unexplored          States   
 00:00:01   0.0 GB     3      12479        208           0 (0 % data)         10083    
--------------------
... Checking statistics:
..... Found 0 bugs.
... Scheduling statistics:
..... Explored 10083 distinct states over 208 timelines
..... Explored 12479 distinct schedules
..... Finished 41 search tasks (0 pending)
..... Number of steps explored: 0 (min), 11 (avg), 30 (max).
... Elapsed 1 seconds and used 0.0 GB
.. Result: correct for any depth
. Done
... Checker run finished.
~~ [PTool]: Thanks for using P! ~~
... Checker run terminated.
````
